### PR TITLE
feat(TimeRelease): Add support for scheduling and executing time-release transfers 

### LINKED ISSUE #2261: Time Release Pallet Cancellable Future Transfers
###  Description
As a user of the Time Release Pallet, I want to be able to schedule a transfer in the future that can still be cancelled by the initiator.

### Acceptance Criteria
- [x] I can create a transfer that will happen at a future block
- [ ] #2275
- [x] Once transferred, it works as it does now.

**Scheduling a Transfer**
- [x] A user can schedule a transfer to execute at a specific future block based on frequency block time.
- [x] The scheduled transfer reserves (holds) the sender’s total scheduled amount until execution or cancellation.
- [x]  The Time Release Pallet schedules the transfer by augmenting the ScheduleOrigin config of the Scheduler Pallet to include EnsureTimeReleaseOrigin.

**Executing a Scheduled Transfer**
- [x] When the scheduled block arrives, the Scheduler automatically executes the transfer by calling execute_scheduled_transfer.

execute_scheduled_transfer must:
- [x] Deduct the transfer amount from the sender’s held (reserved) balance, not their free balance.
- [x] Transfer the amount to the receiver and freeze the receiver’s balance upon receipt, similar to an immediate transfer.

**Access Control & Security**
- [x] Only the Time Release Origin can execute the execute_scheduled_transfer extrinsic.
- [x] No other user or external actor can execute the transfer manually.
- [x] The Scheduler is the only entity authorized to trigger the execution at the designated block time.

### Ideas/Notes
- One option is using the scheduler pallet
- Another option is having a lead in period on the transfer where it can be cancelled still

### PR DESCRIPTION
**Pull request by enddynayn**
Add ability to schedule a time release transfer in the future.

- Users can now schedule transfers to execute at a future block based on frequency block time.  
- Added a time release schedule by augmenting `ScheduleOrigin` in the Scheduler Pallet to include `EnsureTimeReleaseOrigin`, allowing the Time Release Pallet to manage scheduled calls.  
- Scheduled transfers reserve the sender’s total scheduled amount until execution.  
- Implemented `execute_scheduled_transfer`, which transfers funds from the sender’s reserved balance instead of their free balance.  
- Ensured that only the Time Release Origin can execute scheduled transfers, restricting manual execution.  
- Allowed initiators to cancel scheduled transfers before execution.  
- Improved failure handling to ensure transactions fail gracefully if execution conditions are not met.  

Closes #2261 

scheduled-transfer-sequenced-diagram

### CODE REVIEW

File: pallets/time-release/src/lib.rs [hunk 1]
```diff
@@ -90,20 +113,37 @@ pub mod module {
...
 
 		/// Required origin for time-release transfer.
-		type TransferOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+		type TransferOrigin: EnsureOrigin<
+			<Self as frame_system::Config>::RuntimeOrigin,
+			Success = Self::AccountId,
+		>;
+
+		/// Timerelase origin
+		type TimeReleaseOrigin: EnsureOrigin<
```
Comment (enddynayn) [inline pallets/time-release/src/lib.rs:143]: rename: ScheduledTransferOrigin

File: pallets/time-release/src/tests.rs [hunk 2]
```diff
@@ -415,6 +415,230 @@ fn exceeding_maximum_schedules_should_fail() {
...
+			schedule.clone(),
+			when
+		));
+
+		assert_ok!(TimeRelease::execute_scheduled_transfer(
+			Origin::<Test>::TimeRelease(ALICE).into(),
+			BOB,
+			schedule
+		));
+		assert_eq!(PalletBalances::free_balance(BOB), 20u64);
```
Comment (shannonwells) [inline pallets/time-release/src/tests.rs:586]: These tests could use comments reminding a reader how much hold/freeze balance these accounts have.  

Also it seems like if there are 2 periods I would expect 2 scheduled transfers. But I haven't looked at this pallet in a really long time.

File: pallets/time-release/src/lib.rs [hunk 3]
```diff
@@ -79,9 +86,25 @@ pub mod module {
 		TimeReleaseVesting,
 	}
 
+	/// A reason for holding funds
+	/// Create a hold reason for this pallet that is aggregated by 'contract_runtime'.
+	#[pallet::composite_enum]
+	pub enum HoldReason {
+		/// Funds are frozen until
```
Comment (shannonwells) [inline pallets/time-release/src/lib.rs:93]: nit:  finish comment, until when?

File: pallets/time-release/src/lib.rs [hunk 4]
```diff
@@ -113,6 +153,21 @@ pub mod module {
 
 		/// The block-number provider.
 		type BlockNumberProvider: BlockNumberProvider<BlockNumber = BlockNumberFor<Self>>;
+
+		/// The aggregate call type
+		type RuntimeCall: Parameter
+			+ Dispatchable<
+				RuntimeOrigin = <Self as frame_system::Config>::RuntimeOrigin,
+				PostInfo = PostDispatchInfo,
+			> + From<Call<Self>>
```
Comment (shannonwells) [inline pallets/time-release/src/lib.rs:162]: nit: indentation is confusing here

File: pallets/time-release/src/lib.rs [hunk 5]
```diff
@@ -450,6 +587,122 @@ impl<T: Config> Pallet<T> {
...
+		T::Currency::balance_on_hold(&HoldReason::TimeReleaseScheduledVesting.into(), &who)
+	}
+
+	fn ensure_sufficient_hold_balance(from: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+		ensure!(Self::hold_balance_for(&from) >= amount, Error::<T>::InsufficientBalanceToFreeze);
+
+		Ok(())
+	}
+
+	fn validate_and_get_schedule_amount(
```
Comment (shannonwells) [inline pallets/time-release/src/lib.rs:601]: nit: seems like this could be inlined

File: pallets/time-release/src/lib.rs [hunk 6]
```diff
@@ -450,6 +587,122 @@ impl<T: Config> Pallet<T> {
...
+
+		transfer_fn(from, to, schedule_amount)?;
+		Self::update_freeze(to, total_amount)?;
+
+		Self::append_release_schedule(to, schedule)?;
+
+		Ok(())
+	}
+
+	fn finalize_vesting_transfer_from_free_balance(
```
Comment (shannonwells) [inline pallets/time-release/src/lib.rs:691]: nit: These last two abstractions seem unnecessary since they are each used in only one place and the single line in each are clear what they're doing.

File: pallets/time-release/src/lib.rs [hunk 7]
```diff
@@ -268,15 +323,10 @@ pub mod module {
...
 
-			if to == from {
-				ensure!(
-					T::Currency::balance(&from) >=
-						schedule.total_amount().ok_or(ArithmeticError::Overflow)?,
-					Error::<T>::InsufficientBalanceToFreeze,
-				);
-			}
+			let total_amount = schedule.total_amount().ok_or(ArithmeticError::Overflow)?;
+			Self::ensure_suffient_free_balance(&from, &to, total_amount)?;
```
Comment (shannonwells) [inline pallets/time-release/src/lib.rs:327]: spelling for a used function
```suggestion
			Self::ensure_sufficient_free_balance(&from, &to, total_amount)?;
```

File: pallets/time-release/src/lib.rs [hunk 8]
```diff
@@ -450,6 +587,122 @@ impl<T: Config> Pallet<T> {
...
+			schedule_amount,
+			Exact,
+			Free,
+			Polite,
+		)?;
+
+		Ok(())
+	}
+
+	fn ensure_suffient_free_balance(
```
Comment (shannonwells) [inline pallets/time-release/src/lib.rs:659]: spelling for functions should be correct
```suggestion
	fn ensure_sufficient_free_balance(
```

**Review by shannonwells — CHANGES_REQUESTED** [review]
The misspelled function needs to be corrected; everything else is non-blocking, but I do have a strong preference against one line abstractions.  I think it makes the code harder to follow.

File: pallets/time-release/src/tests.rs (hunk 2 above)
Comment (enddynayn) [reply pallets/time-release/src/tests.rs:586]: There is only 1 transfer here.

File: pallets/time-release/src/lib.rs (hunk 4 above)
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:162]: Yeah, I agree, the linter is doing that.

File: pallets/time-release/src/lib.rs (hunk 5 above)
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:601]: I agree `make format` is doing this.

File: pallets/time-release/src/lib.rs (hunk 6 above)
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:691]: I thought this made the code a little easier to read. Let me know if you want me to change it.

File: pallets/time-release/src/lib.rs (hunk 4 above)
Comment (shannonwells) [reply pallets/time-release/src/lib.rs:162]: well poop on that linter

File: pallets/time-release/src/lib.rs [hunk 9]
```diff
@@ -450,6 +587,122 @@ impl<T: Config> Pallet<T> {
...
+	/// * `ArithmeticError::Overflow` - If the transfer amount overflows.
+	/// * `Error::InsufficientBalanceToFreeze` - If the hold balance is insufficient.
+	fn transfer_from_hold_balance(
+		from: &T::AccountId,
+		to: &T::AccountId,
+		schedule_amount: BalanceOf<T>,
+	) -> DispatchResult {
+		// Exact - Will fail if the exact amount cannot be transfered
+		// Free - deposit into the free balance but will locked following
+		// Polite - Execute with regular previlages it consideres existential
```
Comment (JoeCap08055) [inline pallets/time-release/src/lib.rs:645]: I'm a little confused on these comments; perhaps a little more clarification?

File: pallets/time-release/src/lib.rs [hunk 10]
```diff
@@ -326,16 +376,95 @@ pub mod module {
...
+			let total_amount = Self::validate_and_get_schedule_amount(&schedule)?;
+			Self::ensure_sufficient_hold_balance(&from, total_amount)?;
+
+			Self::finalize_vesting_transfer_from_hold(&from, &to, schedule.clone())?;
+
+			Self::deposit_event(Event::ReleaseScheduleAdded {
+				from,
+				to,
+				release_schedule: schedule,
+			});
```
Comment (JoeCap08055) [inline pallets/time-release/src/lib.rs:459]: I'm confused on why we're emitting this event; didn't we emit this event when the schedule was initially added, and now we're executing a previously-added schedule? Or maybe we just have a naming problem...?

File: runtime/common/src/constants.rs [hunk 11]
```diff
@@ -121,8 +121,7 @@ pub type HandleSuffixMax = ConstU16<99>;
 // -end- Handles Pallet
 
 // --- TimeRelease Pallet ---
-// Update
-pub type MinReleaseTransfer = ConstU128<0>;
+pub type MinReleaseTransfer = ConstU128<{ currency::EXISTENTIAL_DEPOSIT }>;
```
Comment (JoeCap08055) [inline runtime/common/src/constants.rs:124]: Hmm, could this be problematic?
I can understand that we might not want to schedule a transfer that might result in an underfunded account that will just get reaped, but won't this also prevent scheduling potentially legitimate transfers to accounts that already hold enough tokens to not be reaped?

Or is this sort of a "lesser of two evils" situation, ie yes, it disallows some minimal (and therefore unlikely?) scheduled transfer amounts, but guarantees that (at least at the time of scheduling) we won't schedule a transfer that will end up getting reaped?

Or, am I misunderstanding the purpose of this value?

Comment (enddynayn) [reply runtime/common/src/constants.rs:124]: This amount is very small, just one cent. We can reasonably expect users to send transfers larger than that. The main purpose of this limit is to help prevent storage spam.

File: pallets/time-release/src/lib.rs (hunk 10 above)
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:459]: Resolution to add documentation.

File: pallets/time-release/src/lib.rs (hunk 9 above)
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:657]: Oops, left over clean up work.

Comment (JoeCap08055) [reply pallets/time-release/src/lib.rs:657]: Suggest either:
* Replace comment with a reference to https://github.com/paritytech/polkadot-sdk/blob/c8d33396345237c1864dfc0a9b2172b7dfe7ac8f/substrate/frame/support/src/traits/tokens/fungible/hold.rs#L346 for the meaning of these parameters
* Replace comment with copied text of above reference

**Review by JoeCap08055 — APPROVED** [review]
one comment nit, otherwise looks great!

File: pallets/time-release/src/lib.rs [hunk 12]
```diff
@@ -326,16 +383,100 @@ pub mod module {
...
+		///
+		/// * [`Event::ReleaseScheduleAdded`] - Indicates that a new release schedule was added.
+		#[pallet::call_index(5)]
+		#[pallet::weight(T::WeightInfo::execute_scheduled_transfer())]
+		pub fn execute_scheduled_transfer(
+			origin: OriginFor<T>,
+			dest: <T::Lookup as StaticLookup>::Source,
+			schedule: ReleaseScheduleOf<T>,
+		) -> DispatchResult {
+			let from = T::TimeReleaseOrigin::ensure_origin(origin)?;
```
Comment (aramikm) [inline pallets/time-release/src/lib.rs:454]: Q: I'm a little confused about the enforcement of the origin. How hard is it to just wrap the origin with `Origin::<T>::TimeRelease` and just call the extrinsic or that is not possible?

File: pallets/time-release/src/lib.rs [hunk 13]
```diff
@@ -376,23 +517,23 @@ impl<T: Config> Pallet<T> {
...
-		Self::update_freeze(&to, total_amount)?;
+		let schedule_call =
+			<T as self::Config>::RuntimeCall::from(Call::<T>::execute_scheduled_transfer {
+				dest,
+				schedule,
+			});
 
-		<ReleaseSchedules<T>>::try_append(to, schedule)
-			.map_err(|_| Error::<T>::MaxReleaseSchedulesExceeded)?;
+		let _ = T::SchedulerProvider::schedule(
```
Comment (aramikm) [inline pallets/time-release/src/lib.rs:532]: It returns () so no need to ignore the result.
```suggestion
		T::SchedulerProvider::schedule(
```

File: runtime/frequency/src/lib.rs [hunk 14]
```diff
@@ -121,7 +123,20 @@ use frame_support::traits::{TryStateSelect, UpgradeCheckSelect};
...
-/// Interface to collective pallet to propose a proposal.
+pub struct SchedulerProvider;
+
+impl SchedulerProviderTrait<RuntimeOrigin, BlockNumber, RuntimeCall> for SchedulerProvider {
+	fn schedule(
+		origin: RuntimeOrigin,
+		when: BlockNumber,
+		call: Box<RuntimeCall>,
+	) -> Result<(), DispatchError> {
+		let _ = Scheduler::schedule(origin, when, None, LOWEST_PRIORITY, call)?;
```
Comment (aramikm) [inline runtime/frequency/src/lib.rs:134]:
```suggestion
		Scheduler::schedule(origin, when, None, LOWEST_PRIORITY, call)?;
```

File: pallets/time-release/src/tests.rs [hunk 15]
```diff
@@ -415,6 +415,230 @@ fn exceeding_maximum_schedules_should_fail() {
 	});
 }
 
+#[test]
+fn schedule_transfer_success() {
```
Comment (aramikm) [inline pallets/time-release/src/tests.rs:419]: if you already didn't have a test it might be useful to a have test that would cause the account to get reaped or any one the edge cases after transfer.

**Review by aramikm — COMMENTED** [review]
Great effort but there are some fixes and questions and nits.

I also see the following warnings when I run the node which are not great and not sure if problematic.

```
warning: patch for `bandersnatch_vrfs` uses the features mechanism. default-features and features will not take effect because the patch dependency does not support this mechanism
warning: Patch `bandersnatch_vrfs v0.0.4 (https://github.com/davxy/ring-vrf.git?rev=48b642a#48b642a0)` was not used in the crate graph.
Perhaps you misspelled the source URL being patched.
Possible URLs for `[patch.<URL>]`:
    https://github.com/w3f/ring-vrf
warning: Patch `fflonk v0.1.1 (https://www.github.com/w3f/fflonk?rev=be95d4c971b1d15b5badfc06ff13f5c07987d484#be95d4c9)` was not used in the crate graph.
Check that the patched package version and available features are compatible

```

File: pallets/time-release/src/lib.rs [hunk 16]
```diff
@@ -326,16 +383,100 @@ pub mod module {
...
+			let from = T::TransferOrigin::ensure_origin(origin)?;
+			let to = T::Lookup::lookup(dest.clone())?;
+
+			let total_amount = Self::validate_and_get_schedule_amount(&schedule)?;
+			Self::ensure_sufficient_free_balance(&from, &to, total_amount)?;
+
+			let _ = Self::update_schedule_vesting_hold(
+				&from,
+				schedule.total_amount().ok_or(ArithmeticError::Overflow)?,
+			);
```
[the remark this answers is not in the document]
Comment (aramikm) [reply pallets/time-release/src/lib.rs:416]: This one is important since it is eating the error

File: Cargo.toml [hunk 17]
```diff
@@ -162,6 +162,12 @@ substrate-frame-rpc-system = { git = "https://github.com/paritytech/polkadot-sdk
 substrate-prometheus-endpoint = { git = "https://github.com/paritytech/polkadot-sdk", branch = "release-polkadot-v1.13.0" }
 sp-debug-derive = { git = "https://github.com/paritytech/polkadot-sdk", branch = "release-polkadot-v1.13.0", default-features = false }
 
+# hotfix for https://github.com/paritytech/polkadot-sdk/issues/7653
+[patch.'https://github.com/w3f/fflonk']
+fflonk = { git = "https://www.github.com/w3f/fflonk", rev = "be95d4c971b1d15b5badfc06ff13f5c07987d484" }
+bandersnatch_vrfs = { git = "https://github.com/davxy/ring-vrf.git", rev = "48b642a", default-features = false }
```
[the remark this answers is not in the document]
Comment (enddynayn) [reply Cargo.toml:168]: This is currently an issue on main.

File: pallets/time-release/src/lib.rs [hunk 18]
```diff
@@ -450,6 +599,106 @@ impl<T: Config> Pallet<T> {
...
+		Ok(())
+	}
+
+	fn ensure_sufficient_free_balance(
+		from: &T::AccountId,
+		to: &T::AccountId,
+		amount: BalanceOf<T>,
+	) -> DispatchResult {
+		if to == from {
+			ensure!(T::Currency::balance(&from) >= amount, Error::<T>::InsufficientBalanceToFreeze,);
```
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:688]: Perhaps, but this might be a bit out of scope. These errors weren’t created by me; they are part of the main setup. If we’d like to make changes to the current setup, I’d suggest creating a new ticket for it. Let me know what you think!

File: pallets/time-release/src/lib.rs [hunk 19]
```diff
@@ -450,6 +599,106 @@ impl<T: Config> Pallet<T> {
...
+		Ok(())
+	}
+
+	fn validate_and_get_schedule_amount(
+		schedule: &ReleaseScheduleOf<T>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		ensure_valid_release_schedule::<T>(schedule)
+	}
+
+	fn append_release_schedule(
```
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:619]: I think having these go into functions makes it easier for name changes and updates.

File: pallets/time-release/src/lib.rs [hunk 20]
```diff
@@ -433,6 +574,14 @@ impl<T: Config> Pallet<T> {
 		Ok(())
 	}
 
+	fn update_schedule_vesting_hold(
```
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:577]: I see your point, but I think isolating dependencies from the outside can be beneficial. Since names and structures might change over time, encapsulating them could make the code more maintainable and easier to work with in the long run. Let me know what you think!

File: pallets/time-release/src/lib.rs (hunk 12 above)
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:454]: Good question! I believe that’s not possible. The Origin serves as an access control mechanism. We ask the same question for any origin that requires sudo or governance, and if it were possible to bypass it, those mechanisms would break.

File: pallets/time-release/src/lib.rs [hunk 21]
```diff
@@ -450,6 +599,106 @@ impl<T: Config> Pallet<T> {
...
+		T::Currency::balance_on_hold(&HoldReason::TimeReleaseScheduledVesting.into(), &who)
+	}
+
+	fn ensure_sufficient_hold_balance(from: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+		ensure!(Self::hold_balance_for(&from) >= amount, Error::<T>::InsufficientBalanceToFreeze);
+
+		Ok(())
+	}
+
+	fn validate_and_get_schedule_amount(
```
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:613]: I wanted to avoid modifying the existing code since it seems to be outside the scope of this pallet. However, you make a great point, and we could create a ticket to further improve it. I also have a few changes in mind that could make it much easier to work with.

File: pallets/time-release/src/lib.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (aramikm) [reply pallets/time-release/src/lib.rs:688]: This is not a big deal but since they recommend using `T::Currency::reducible_balance` instead of `T::Currency::balance` it might worth taking a look at it.

File: pallets/time-release/src/lib.rs (hunk 12 above)
Comment (aramikm) [reply pallets/time-release/src/lib.rs:454]: I might be wrong but I think sudo or governance check it against a state (accounts) stored on-chain but this doesn't seems to do that.

Comment (enddynayn) [reply pallets/time-release/src/lib.rs:454]: That’s true. Let me clarify my understanding. The origin is set by the runtime/node and is typically one of the following:
```
enum RawOrigin {
    Signed,
    Root,
    None,
}
```

The origin type is determined based on who or what initiated the action:
- Signed -> A user-submitted transaction.
- Root -> A governance or privileged system call.
- None -> A system-initiated operation without a specific identity.

In this case, when a signed transaction comes our way we modify the origin and pass it to the scheduler. Only in this piece of code can you get that origin (unless we add it else where). 

A transaction **cannot modify its own origin**, as origin assignment is strictly handled internally by the runtime for security and access control reasons.

Let me know if you have any questions! This is my understanding, but I might have missed something. Happy to discuss further.

File: pallets/time-release/src/lib.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/time-release/src/lib.rs:688]: I think thats a good idea. I will create a new ticket for this.

File: Cargo.toml [hunk 22]
```diff
@@ -162,6 +162,12 @@ substrate-frame-rpc-system = { git = "https://github.com/paritytech/polkadot-sdk
 substrate-prometheus-endpoint = { git = "https://github.com/paritytech/polkadot-sdk", branch = "release-polkadot-v1.13.0" }
 sp-debug-derive = { git = "https://github.com/paritytech/polkadot-sdk", branch = "release-polkadot-v1.13.0", default-features = false }
 
+# hotfix for https://github.com/paritytech/polkadot-sdk/issues/7653
```
Comment (enddynayn) [inline Cargo.toml:165]: Remove and comment out Cargo CI and create a ticket too fix. Issue should resolve upon upgrade to new Polkadot version.

File: pallets/time-release/src/lib.rs [hunk 23]
```diff
@@ -326,16 +383,100 @@ pub mod module {
...
+
+		/// This function schedules a transfer by calling the `Scheduler` pallet's `schedule` function.
+		/// The transfer will be executed at the specified block number.
+		///
+		/// * [`Error::InsufficientBalanceToFreeze`] - Insufficient amount of balance to freeze.
+		/// * [`Error::MaxReleaseSchedulesExceeded`] - Failed because the maximum release schedules was exceeded-
+		/// * [`ArithmeticError::Overflow] - Failed because of an overflow.
+		///
+		#[pallet::call_index(4)]
+		#[pallet::weight(T::WeightInfo::schedule_transfer().saturating_add(T::WeightInfo::transfer()))]
```
Comment (enddynayn) [inline pallets/time-release/src/lib.rs:400]: execute_schedule_transfer
